sql/src/main/java/com/vmware/ddlog/translator/SqlSemantics.java

```diff
@@ -0,0 +1,175 @@
+package com.vmware.ddlog.translator;
+
+import com.vmware.ddlog.ir.*;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+
+/**
+ * This is a singleton pattern class
+ * encoding various properties about SQL semantics.
+ */
+public class SqlSemantics {
+    private final HashSet<String> aggregateFunctions = new HashSet<String>();
+    private final HashMap<String, DDlogEBinOp.BOp> arithmeticFunctions = new HashMap<String, DDlogEBinOp.BOp>();
+    private final HashMap<String, DDlogEBinOp.BOp> stringFunctions = new HashMap<String, DDlogEBinOp.BOp>();
+    private final HashMap<String, DDlogEBinOp.BOp> booleanFunctions = new HashMap<String, DDlogEBinOp.BOp>();
+
+    private SqlSemantics() {
+        this.aggregateFunctions.add("count");
+        this.aggregateFunctions.add("sum");
+        this.aggregateFunctions.add("avg");
+        this.aggregateFunctions.add("min");
+        this.aggregateFunctions.add("max");
+
+        this.arithmeticFunctions.put("a_eq", DDlogEBinOp.BOp.Eq);
+        this.arithmeticFunctions.put("a_neq", DDlogEBinOp.BOp.Neq);
+        this.arithmeticFunctions.put("a_lt", DDlogEBinOp.BOp.Lt);
+        this.arithmeticFunctions.put("a_gt", DDlogEBinOp.BOp.Gt);
+        this.arithmeticFunctions.put("a_lte", DDlogEBinOp.BOp.Lte);
+        this.arithmeticFunctions.put("a_gte", DDlogEBinOp.BOp.Gte);
+        this.arithmeticFunctions.put("a_plus", DDlogEBinOp.BOp.Plus);
+        this.arithmeticFunctions.put("a_minus", DDlogEBinOp.BOp.Minus);
+        this.arithmeticFunctions.put("a_mod", DDlogEBinOp.BOp.Mod);
+        this.arithmeticFunctions.put("a_times", DDlogEBinOp.BOp.Times);
+        this.arithmeticFunctions.put("a_div", DDlogEBinOp.BOp.Div);
+        // The following don't work for 64 bits
+        // this.arithmeticFunctions.put("a_shiftr", DDlogEBinOp.BOp.ShiftR);
+        // this.arithmeticFunctions.put("a_shiftl", DDlogEBinOp.BOp.ShiftL);
+        this.arithmeticFunctions.put("a_band", DDlogEBinOp.BOp.BAnd);
+        this.arithmeticFunctions.put("a_bor", DDlogEBinOp.BOp.BOr);
+        this.arithmeticFunctions.put("a_bxor", DDlogEBinOp.BOp.BXor);
+
+        this.stringFunctions.put("s_concat", DDlogEBinOp.BOp.Concat);
+        this.stringFunctions.put("s_eq", DDlogEBinOp.BOp.Eq);
+        this.stringFunctions.put("s_neq", DDlogEBinOp.BOp.Neq);
+
+        this.booleanFunctions.put("b_eq", DDlogEBinOp.BOp.Eq);
+        this.booleanFunctions.put("b_neq", DDlogEBinOp.BOp.Neq);
+        this.booleanFunctions.put("b_and", DDlogEBinOp.BOp.And);
+        this.booleanFunctions.put("b_or", DDlogEBinOp.BOp.Or);
+        this.booleanFunctions.put("b_impl", DDlogEBinOp.BOp.Impl);
+    }
+
+    public static SqlSemantics semantics = new SqlSemantics();
```

Comment:
Why does this need an instance? As opposed to calls like isAggregateFunction() etc being static methods?